truffle/src/main/ruby/core/rubinius/common/encoding.rb

```diff
@@ -0,0 +1,614 @@
+          end
+
+          if source_encoding_name.to_sym == @source_encoding.name and
+              destination_encoding_name.to_sym == @destination_encoding.name
+            msg = "#{error_bytes_msg} from #{source_encoding_name} to #{destination_encoding_name}"
+          else
+            msg = "#{error_bytes_msg} to #{destination_encoding_name} in conversion from #{source_encoding_name}"
+            transcoder = @converters.first
+            msg << " to #{transcoder.target}"
+          end
+
+          exc = UndefinedConversionError.new msg
+      end
+
+      Rubinius.privately do
+        exc.source_encoding_name = source_encoding_name
+        src = Rubinius::Type.try_convert_to_encoding source_encoding_name
+        exc.source_encoding = src unless undefined.equal? src
+
+        exc.destination_encoding_name = destination_encoding_name
+        dst = Rubinius::Type.try_convert_to_encoding destination_encoding_name
+        exc.destination_encoding = dst unless undefined.equal? dst
+
+        if error_char
+          error_char.force_encoding src unless undefined.equal? src
+          exc.error_char = error_char
+        end
+
+        if result == :invalid_byte_sequence or result == :incomplete_input
+          exc.error_bytes = error_bytes.force_encoding Encoding::ASCII_8BIT
+
+          if bytes = error[:read_again_bytes]
+            exc.readagain_bytes = bytes.force_encoding Encoding::ASCII_8BIT
+          end
+        end
+
+        if result == :invalid_byte_sequence
+          exc.incomplete_input = false
+        elsif result == :incomplete_input
+          exc.incomplete_input = true
+        end
+      end
+
+      exc
+    end
+
+    def primitive_errinfo
+      Rubinius.primitive :encoding_converter_primitive_errinfo
+      raise PrimitiveFailure, "Encoding::Converter#primitive_errinfo primitive failed"
+    end
+
+    def convpath
+      path = []
+      a = 0
+      b = @convpath.size - 1
+
+      while a < b
+        path << [Encoding.find(@convpath[a].to_s), Encoding.find(@convpath[a + 1].to_s)]
+        a += 1
+      end
+
+      path << "xml_text_escape" if @options & XML_TEXT_DECORATOR != 0
+      path << "xml_attr_content_escape" if @options & XML_ATTR_CONTENT_DECORATOR != 0
+      path << "xml_attr_quote" if @options & XML_ATTR_QUOTE_DECORATOR != 0
+      path << "universal_newline" if @options & UNIVERSAL_NEWLINE_DECORATOR != 0
+      path << "crlf_newline" if @options & CRLF_NEWLINE_DECORATOR != 0
+      path << "cr_newline" if @options & CR_NEWLINE_DECORATOR != 0
+
+      path
+    end
+
+    def inspect
+      "#<Encoding::Converter: #{source_encoding.name} to #{destination_encoding.name}>"
+    end
+
+    def replacement=(str)
+      str = StringValue(str)
+
+      @replacement = str.encode(@destination_encoding)
+    end
+
+    class TranscodingPath
+      @paths = {}
+      @load_cache = true
+      @cache_valid = false
+      @transcoders_count = TranscodingMap.size
```

Comment:
I was trying to avoid modifying their source altogether since it would make future updates easier.  But I'm open to other options.